compiler/src/main/kotlin/ast/ast.kt

```diff
@@ -74,39 +122,88 @@ data class NegativeExpr(val ctx: CellmataParser.NegativeExprContext, val value:
 
-data class ArrayLookupExpr(val ctx: CellmataParser.ArrayLookupExprContext, val ident: String = MAGIC_UNDEFINED_STRING, val index: Expr) : Expr()
+data class ArrayLookupExpr(
+    val ctx: CellmataParser.ArrayLookupExprContext,
+    val ident: String = MAGIC_UNDEFINED_STRING,
+    val index: Expr
+) : Expr()
 
-data class ArrayBodyExpr(val ctx: CellmataParser.ArrayValueExprContext, val values: List<Expr>) : Expr()
+/**
+ * @param ctx Node in the parse tree this node corresponds to.
+ * @param values Elements of the array.
+ * @param declaredType Type listed before the body/values of the array.
+ */
+data class ArrayBodyExpr(
+    val ctx: CellmataParser.ArrayValueExprContext,
+    val values: List<Expr>,
+    val declaredType: Type? = UncheckedType
+) : Expr()
 
 data class ParenExpr(val ctx: CellmataParser.ParenExprContext, val expr: Expr) : Expr()
 
 data class NamedExpr(val ctx: ParseTree, var ident: String = MAGIC_UNDEFINED_STRING) : Expr()
 
 data class ModuloExpr(val ctx: CellmataParser.ModuloExprContext, val left: Expr, val right: Expr) : Expr()
 
-data class FuncExpr(val ctx: CellmataParser.FuncExprContext, val args: List<Expr>, var ident: String = MAGIC_UNDEFINED_STRING) : Expr()
+/**
+ * Presents a function call in abstract syntax tree
+ */
+data class FuncExpr(
+    val ctx: CellmataParser.FuncExprContext,
+    val args: List<Expr>,
+    var ident: String = MAGIC_UNDEFINED_STRING
+) : Expr()
 
-data class StateIndexExpr(val ctx: CellmataParser.StateIndexExprContext) : Expr()
+data class StateIndexExpr(val ctx: CellmataParser.StateIndexExprContext) : Expr(type = IntegerType)
 
 // Literals
-data class IntLiteral(val ctx: CellmataParser.Integer_literalContext, var value: Int = -1) : Expr()
+/**
+ * Represents a concrete integer value in the abstract syntax tree
+ */
+data class IntLiteral(val ctx: CellmataParser.Integer_literalContext, var value: Int = -1) : Expr(type = IntegerType)
 
-data class BoolLiteral(val ctx: CellmataParser.Bool_literalContext, var value: Boolean = false) : Expr()
+/**
+ * Represents a concrete boolean value in the abstract syntax tree
+ */
+data class BoolLiteral(val ctx: CellmataParser.Bool_literalContext, var value: Boolean = false) : Expr(type = BooleanType)
 
-data class FloatLiteral(val ctx: CellmataParser.Float_literalContext, var value: Float = 0.0F): Expr()
+/**
+ * Represent a concrete float value in the abstract syntax tree
+ */
+data class FloatLiteral(val ctx: CellmataParser.Float_literalContext, var value: Float = 0.0F) : Expr(type = FloatType)
 
 // Type conversion
+/**
+ * Internal node used to present an implicit conversion from integer to float.
+ */
 data class IntToFloatConversion(val expr: Expr) : Expr()
 
+/**
+ * Internal node used to represent an implicit conversion from a array of states to a neighbourhood.
+ */
 data class StateArrayToActualNeighbourhoodConversion(val expr: Expr) : Expr()
 
 /*
  * Declarations
```

Comment:
This should be an ordinary comment `//`, as I assume it refers to the category of the following definitions.

Reply:
No, the "comment" is the documentation for `StateArrayToActualNeighbourhoodConversion`